man8/cache_metadata_size.8
.TH CACHE_METADATA_SIZE 8 "Thin Provisioning Tools" "Red Hat, Inc."
.SH NAME
cache_metadata_size \- Estimate the size of the metadata device needed for a given configuration

.SH SYNOPSIS
.B cache_metadata_size
.RB [ options ]

.SH DESCRIPTION
.B cache_metadata_size
estimates the size of the metadata.  Either \-\-nr\-blocks, or \-\-block\-size
and \-\-device\-size must be specified.

.SH OPTIONS
.IP "\fB\-\-nr\-blocks\fP \<natural\>"
Specify the number of cache blocks.

.IP "\fB\-\-block-size\fP \<sectors\>"
Specify the size of each cache block in 512 byte sectors.

.IP "\fB\-\-device\-size\fP \<sectors\>"
Specify total size of the fast device used in the cache.  In 512 byte sectors.

.IP "\fB\-\-max\-hint\-width\fP \<nr bytes\>"
Cache policies use a per block 'hint' to record extra info (for instance hit
counts).  At the moment all policies use a 4 byte hint width.  If you want to use
a different hint width specify it with this switch.

.SH EXAMPLE
.B cache_metadata_size --nr-blocks 10240
.sp
.B cache_metadata_size --block-size 128 --device-size 1024000

.SH SEE ALSO
.B cache_dump(8)
.B cache_repair(8)
.B cache_restore(8)

.SH AUTHOR
Joe Thornber <ejt@redhat.com>
.br
Heinz Mauelshagen <heinzm@redhat.com>



CHANGES
v0.6.2
======

- Fix bug in thin_delta
- Fix recent regression in thin_repair.
- Force g++-98 dialect
- Fix bug in thin_trim

v0.6.1
======

Fix bug in thin_dump when using metadata snaps.

v0.6.0
======

thin_ls

v0.5.6
======

era_invalidate may be run on live metadata if the --metadata-snap option is given.

v0.5.5
======

You may now give the --metadata_snap option to thin_delta without specifying where the snap is.

Update man pages to make it clearer that most tools shouldn't be run on live metadata.

Fix some bugs in the metadata reference counting for thin_check.

v0.5.1
======

Fix bug where the tools would crash if given a very large metadata
device to restore to.

v0.5
====

- thin_delta, thin_trim
- --clear-needs-check flag for cache_check
- space map checking for thin check

v0.4
====

- All tools switch to using libaio.  This gives a large performance
  boost, especially to the write focused tools like thin_restore.

- Added a progress monitor to thin_restore, cache_restore and era_restore

- Added a --quiet/-q option to *_restore to turn off the progress bar

- Removed variable hint size support from cache tools.  The kernel
  still only supports a fixed 32bit width.  This will have a side
  effect of reducing the executable sizes due to less template
  instatiation.

- Tools rolled into a single executable to save space.

- Fixed some bugs when walking bitsets (possibly effecting cache_dump
  and cache_check).

caching/cache_metadata_size.cc
#include "version.h"

#include "caching/commands.h"

#include <boost/lexical_cast.hpp>
#include <boost/optional.hpp>
#include <iostream>
#include <getopt.h>
#include <stdint.h>
#include <stdexcept>

using namespace caching;
using namespace std;

//----------------------------------------------------------------

cache_metadata_size_cmd::flags::flags()
  : max_hint_width(4) {

	// Dance around some spurious compiler warnings
	device_size = 0;
	block_size = 0;
	nr_blocks = 0;

	device_size.reset();
	block_size.reset();
	nr_blocks.reset();
}

cache_metadata_size_cmd::parse_result
cache_metadata_size_cmd::parse_command_line(string const &prog_name, int argc, char **argv, flags &fs)
{
	int c;
	char const short_opts[] = "hV";
	option const long_opts[] = {
		{ "block-size", required_argument, NULL, 0 },
		{ "device-size", required_argument, NULL, 1 },
		{ "nr-blocks", required_argument, NULL, 2 },
		{ "max-hint-width", required_argument, NULL, 3 },
		{ "help", no_argument, NULL, 'h' },
		{ "version", no_argument, NULL, 'V' },
		{ NULL, no_argument, NULL, 0 }
	};

	while ((c = getopt_long(argc, argv, short_opts, long_opts, NULL)) != -1) {
		switch (c) {
		case 0:
			fs.block_size = boost::lexical_cast<uint32_t>(optarg);
			break;

		case 1:
			fs.device_size = boost::lexical_cast<uint64_t>(optarg);
			break;

		case 2:
			fs.nr_blocks = boost::lexical_cast<uint64_t>(optarg);
			break;

		case 3:
			fs.max_hint_width = boost::lexical_cast<uint32_t>(optarg);
			break;

		case 'h':
			usage(cout);
			return FINISH;
			break;

		case 'V':
			cout << THIN_PROVISIONING_TOOLS_VERSION << endl;
			return FINISH;
			break;

		default:
			usage(cerr);
			throw runtime_error("Invalid command line");
			break;
		}
	}

	return CONTINUE;
}

uint64_t
cache_metadata_size_cmd::get_nr_blocks(flags &fs)
{
	if (fs.device_size) {
		if (!fs.block_size)
			throw runtime_error("If you specify --device-size you must also give --block-size.");

		uint64_t nr_blocks = *fs.device_size / *fs.block_size;
		if (fs.nr_blocks) {
			if (nr_blocks != *fs.nr_blocks)
				throw runtime_error(
					"Contradictory arguments given, --nr-blocks doesn't match the --device-size and --block-size.");
		}

		return nr_blocks;
	}

	if (fs.block_size && !fs.device_size)
		throw runtime_error("If you specify --block-size you must also give --device-size.");

	if (fs.nr_blocks)
		return *fs.nr_blocks;

	throw runtime_error("Please specify either --device-size and --block-size, or --nr-blocks.");
}

uint64_t
cache_metadata_size_cmd::meg(uint64_t n)
{
	return n * 2048;
}

uint64_t
cache_metadata_size_cmd::calc_size(uint64_t nr_blocks, uint32_t max_hint_width)
{
	uint64_t const SECTOR_SIZE = 512;
	uint64_t const TRANSACTION_OVERHEAD = meg(4);
	uint64_t const BYTES_PER_BLOCK = 16;
	uint64_t const HINT_OVERHEAD_PER_BLOCK = 8;

	uint64_t mapping_size = (nr_blocks * BYTES_PER_BLOCK) / SECTOR_SIZE;
	uint64_t hint_size = (nr_blocks * (max_hint_width + HINT_OVERHEAD_PER_BLOCK)) / SECTOR_SIZE;
	return TRANSACTION_OVERHEAD + mapping_size + hint_size;
}

//----------------------------------------------------------------

cache_metadata_size_cmd::cache_metadata_size_cmd()
	: command("cache_metadata_size")
{
}

void
cache_metadata_size_cmd::usage(ostream &out) const
{
	out << "Usage: " << get_name() << " [options]" << endl
	    << "Options:" << endl
	    << "  {-h|--help}" << endl
	    << "  {-V|--version}" << endl
	    << "  {--block-size <sectors>}" << endl
	    << "  {--device-size <sectors>}" << endl
	    << "  {--nr-blocks <natural>}" << endl
	    << "  {--max-hint-width <nr bytes>}" << endl << endl
	    << "These all relate to the size of the fast device (eg, SSD), rather" << endl
	    << "than the whole cached device." << endl;
}

int
cache_metadata_size_cmd::run(int argc, char **argv)
{
	flags fs;

	try {
		switch (parse_command_line(argv[0], argc, argv, fs)) {
		case FINISH:
			return 0;

		case CONTINUE:
			break;
		}

		uint64_t nr_blocks = get_nr_blocks(fs);
		cout << calc_size(nr_blocks, fs.max_hint_width) << " sectors" << endl;

	} catch (std::exception const &e) {
		cerr << e.what() << "\n";
		return 1;
	}

	return 0;
}

//----------------------------------------------------------------
